server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CounterResource.java

```diff
@@ -0,0 +1,253 @@
+
+import static org.jboss.as.controller.PathAddress.pathAddress;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP_ADDR;
+
+import java.util.Optional;
+
+import org.infinispan.counter.api.CounterManager;
+import org.infinispan.counter.api.StrongCounter;
+import org.infinispan.counter.api.WeakCounter;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.server.infinispan.SecurityActions;
+import org.jboss.as.controller.AbstractAddStepHandler;
+import org.jboss.as.controller.AttributeDefinition;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationDefinition;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.OperationStepHandler;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PathElement;
+import org.jboss.as.controller.ReloadRequiredWriteAttributeHandler;
+import org.jboss.as.controller.SimpleAttributeDefinitionBuilder;
+import org.jboss.as.controller.SimpleOperationDefinitionBuilder;
+import org.jboss.as.controller.SimpleResourceDefinition;
+import org.jboss.as.controller.descriptions.ResourceDescriptionResolver;
+import org.jboss.as.controller.registry.AttributeAccess;
+import org.jboss.as.controller.registry.ManagementResourceRegistration;
+import org.jboss.as.controller.services.path.ResolvePathHandler;
+import org.jboss.dmr.ModelNode;
+import org.jboss.dmr.ModelType;
+
+/**
+ * Resource description for the addressable resource
+ * /subsystem=infinispan/cache-container=X/counter=*
+ *
+ * @author Pedro Ruivo
+ * @author Vladimir Blagojevic
+ * @since 9.2
+ */
+public class CounterResource extends SimpleResourceDefinition {
+
+    //atributes
+    static final AttributeDefinition CONFIGURATION = new SimpleAttributeDefinitionBuilder(ModelKeys.CONFIGURATION,
+            ModelType.STRING, false)
+            .setXmlName(Attribute.CONFIGURATION.getLocalName())
+            .setAllowExpression(false)
+            .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+            .build();
+
+    static final AttributeDefinition[] ATTRIBUTES = { CONFIGURATION };
+
+    // operations
+
+    private static final OperationDefinition COUNTER_RESET = buildOperation("counter-reset").build();
+    private static final OperationDefinition COUNTER_INCREASE = buildOperation("counter-increase").build();
+    private static final OperationDefinition COUNTER_DECREASE = buildOperation("counter-decrease").build();
+
+    private final boolean runtimeRegistration;
+    private final ResolvePathHandler resolvePathHandler;
+
+    public CounterResource(PathElement pathElement, ResourceDescriptionResolver descriptionResolver,
+            ResolvePathHandler resolvePathHandler, AbstractAddStepHandler addHandler,
+            OperationStepHandler removeHandler, boolean runtimeRegistration) {
+        super(pathElement, descriptionResolver, addHandler, removeHandler);
```

Comment:
The calls to `CounterResource` in both `Weak/StrongCounterResource` are exactly the same except for pathElement, so you can use the following constructor instead:

```java
public CounterResource(PathElement pathElement, ResourceDescriptionResolver descriptionResolver,
                           ResolvePathHandler resolvePathHandler, boolean runtimeRegistration) {
        super(pathElement, descriptionResolver, new CounterAddHandler(), new CounterRemoveHandler());
        this.resolvePathHandler = resolvePathHandler;
        this.runtimeRegistration = runtimeRegistration;
    }
```

Reply:
Or you could go further and remove `Weak/StrongCounterResource` entirely and just create two `CounterResource` instances with the desired paths in `CacheContainerCountersResource`